### ISSUE TITLE
Pluggable API for third-party geospatial formats

### ISSUE THREAD

**Issue by mourner** [mapbox/mapbox-gl-js#929]
Currently we have awesome GeoJSON support that can handle big amounts of data, but all the work happens on the worker side. This means that to support third-party formats like Geobuf and Shapefile well, we can't just convert to GeoJSON in the app code and pass it to the API, because:

1. The main thread is blocked while the data is converted.
2. Transfering big GeoJSON to the worker is also blocking and extremely expensive because essentially you transfer stringified JSON with memory copy and then parse back.

We need to find a way to plug into the worker to do all the processing there, without the need to compile the decoding code into the bundle.

One way we could approach this is utilizing `importScripts` — we could pass a JS file URL with the decoding code (maybe wrapped and bundled specially for Mapbox GL JS) to the worker, and the worker would in turn load it and use it to process a particular third-party source:

```js
var geobuf = new GeoJSONSource({url: 'mydata.geobuf', decoder: 'mapbox-gl-geobuf.js'});
```

cc @tmcw @jfirebaugh

**Comment by tmcw** [comment]
:+1: Though I'd assume that we would generalize `GeoJSONSource` to be something like `SingleFileSource` in this case.

I do have mild concerns about using importScripts for this in the context of the browserify-bundle issues that have cropped up before, and I'm interested in [workerify](https://github.com/shama/workerify)-like approaches, but... :shrug: those things can happen in prototyping

**Comment by anandthakker** [comment]
I'd be into taking a crack at this for a use case I'm currently trying to deal with (involving using TopoJSON).

Is does the approach outlined here still seem viable?  I.e., generalizing `GeoJSONSource` to `SingleFileSource` (`VectorFileSource`?) that optionally accepts a url to a js decoder that preprocesses whatever comes back from `source.url`.

Is CSP relevant here?

**Comment by jfirebaugh** [comment]
I'm pretty :-1: on this. Yes, CSP is relevant. Since [`importScripts`](https://developer.mozilla.org/en-US/docs/Web/API/WorkerGlobalScope/importScripts) can only load same-domain resources, this would have to be an XHR + eval, AKA an API to inject arbitrary code into your webpage, which is a pretty big security footgun. Plus global eval means you can't do anything like `var decode = require('mapbox-gl-geobuf.js'); process(decode(geobuf));`. Instead we'd have to invent a naming convention that the decoders have to follow, juggle global variables, ensure there are no collisions, etc.

**Comment by tmcw** [comment]
@jfirebaugh okay, so the alternative is to do format conversion in the main thread? Seems fine to me.

**Comment by anandthakker** [comment]
@jfirebaugh how sure are you that those issues are show stoppers?    I ask because it feels like such a bummer if they are, since it means that anything happening on the worker side can't be extended with third party code/plugins.

Potential responses to the issues you raised:

> Instead we'd have to invent a naming convention that the decoders have to follow...

Since you're sending to the worker the url for the third-party script anyway, you could just as easily also send the name of the decoder function.  Or, cleaner, expect that the decoder is a module built with `browserify -r ./index.js:my-plugin-name`, and then send something like `{ url: "...", module: "my-plugin-name" }`.

> juggle global variables, ensure there are no collisions, etc

I think the `require()` approach helps a bit with this, although it's not perfect.  Looks like `vm.runInNewContext()` might be an option: https://github.com/substack/vm-browserify ... even if it does use an iframe under the hood--yuck.

> Since importScripts can only load same-domain resources

Maybe it the same-domain restriction is okay: it's a pretty advanced use case, so having people bundle/serve up the plugin code and gl-js from the same place instead of relying on Mapbox's CDN version seems reasonable.  Also, yall could host "official" plugins on the CDN as well so that they'd be available for the less advanced...

**Comment by anandthakker** [comment]
I've come across a couple of new use cases for some kind of pluggability into workers/sources:

1. Offline caching -- A source implementation that proxies through to vector tile source, but caches to indexedDB (or something)
2. Joining data properties dynamically -- It would be great to be able to use a tileset with just the geometries and an id, and then use the ids to join property data from a different source.  Really useful for  data-driven styling in cases where the properties reflect dynamic or live data (maybe coming from an API or user actions).

Any new thoughts on this topic in general?

**Comment by anandthakker** [comment]
@lucaswoj the core of this enhancement is now implemented via #2667 .  What final items do you think ought to happen before closing this ticket?  Full vetting and adding to the public, documented API?  Just a couple of working examples?  Something in between?

**Comment by mourner** [comment]
@anandthakker opened a new ticket for that #2920
